lib/pryin/interaction_store.ex

```diff
@@ -133,86 +150,97 @@ defmodule PryIn.InteractionStore do
-      interaction ->
-        ecto_query = Interaction.EctoQuery.new(data)
-        interaction = Map.update!(interaction, :ecto_queries, &([ecto_query | &1]))
-        running_interactions = %{state.running_interactions | pid => interaction}
-        {:noreply, %{state | running_interactions: running_interactions}}
-    end
+    state = update_in(state.running_interactions[parent_pid(state, pid)].interaction,
+      fn
+        nil -> nil
+        interaction ->
+          ecto_query = Interaction.EctoQuery.new(data)
+          Map.update!(interaction, :ecto_queries, &([ecto_query | &1]))
+      end)
+    {:noreply, state}
   end
 
   def handle_cast({:add_view_rendering, pid, data}, state) do
-    interaction = Map.get(state.running_interactions, pid)
-    case interaction do
-      nil ->
-        {:noreply, state}
-      interaction ->
-        view_rendering = Interaction.ViewRendering.new(data)
-        interaction = Map.update!(interaction, :view_renderings, &([view_rendering | &1]))
-        running_interactions = %{state.running_interactions | pid => interaction}
-        {:noreply, %{state | running_interactions: running_interactions}}
-    end
+    state = update_in(state.running_interactions[parent_pid(state, pid)].interaction,
+      fn
+        nil -> nil
+        interaction ->
+          view_rendering = Interaction.ViewRendering.new(data)
+          Map.update!(interaction, :view_renderings, &([view_rendering | &1]))
+      end)
+    {:noreply, state}
   end
 
   def handle_cast({:add_custom_metric, pid, data}, state) do
-    interaction = Map.get(state.running_interactions, pid)
-    case interaction do
-      nil ->
-        {:noreply, state}
-      interaction ->
-        custom_metric = Interaction.CustomMetric.new(data)
-        interaction = Map.update!(interaction, :custom_metrics, &([custom_metric | &1]))
-        running_interactions = %{state.running_interactions | pid => interaction}
-        {:noreply, %{state | running_interactions: running_interactions}}
-    end
+    state = update_in(state.running_interactions[parent_pid(state, pid)].interaction,
+      fn
+        nil -> nil
+        interaction ->
+          custom_metric = Interaction.CustomMetric.new(data)
+          Map.update!(interaction, :custom_metrics, &([custom_metric | &1]))
+      end)
+    {:noreply, state}
   end
 
   def handle_cast({:finish_interaction, pid}, state) do
-    {monitor_ref, remaining_monitor_refs} = Map.pop(state.monitor_refs, pid)
-    Process.demonitor(monitor_ref)
-    {finished_interaction, running_interactions} = Map.pop(state.running_interactions, pid)
-    finished_interaction = Map.update!(finished_interaction, :start_time, &trunc(&1 / 1000))
+    {finished_running_interaction, running_interactions} = Map.pop(state.running_interactions, pid)
+    Process.demonitor(finished_running_interaction.ref)
+
+    running_interactions = clear_children(running_interactions, finished_running_interaction.child_pids)
```

Comment:
finish parent process instead, to allow calling finish from joined child processes